src/CustomElements/v1/CustomElements.js

```diff
@@ -0,0 +1,597 @@
+        root.__observer = null;
+        if (this.enableFlush) {
+          this._observers.delete(root.__observer);
+        }
+      }
+    },
+
+    /**
+     * @private
+     */
+    _handleMutations: function(mutations) {
+      for (var i = 0; i < mutations.length; i++) {
+        var mutation = mutations[i];
+        if (mutation.type === 'childList') {
+          // Note: we can't get an ordering between additions and removals, and
+          // so might diverge from spec reaction ordering
+          this._addNodes(mutation.addedNodes);
+          this._removeNodes(mutation.removedNodes);
+        }
+      }
+    },
+
+    /**
+     * @param {NodeList} nodeList
+     * @private
+     */
+    _addNodes: function(nodeList) {
+      for (var i = 0; i < nodeList.length; i++) {
+        var root = nodeList[i];
+
+        if (!isElement(root)) {
+          continue;
+        }
+
+        // Since we're adding this node to an observed tree, we can unobserve
+        this._unobserveRoot(root);
+
+        var walker = createTreeWalker(root);
+        do {
+          var node = /** @type {HTMLElement} */ (walker.currentNode);
+          var definition = this._definitions.get(node.localName);
+          if (definition) {
+            if (!node.__upgraded) {
+              this._upgradeElement(node, definition, true);
+            }
+            if (node.__upgraded && !node.__attached) {
+              node.__attached = true;
+              if (definition && definition.connectedCallback) {
+                definition.connectedCallback.call(node);
+              }
+            }
+          }
+          if (node.shadowRoot) {
+            // TODO(justinfagnani): do we need to check that the shadowRoot
+            // is observed?
+            this._addNodes(node.shadowRoot.childNodes);
+          }
+          if (node.tagName === 'LINK') {
+            var onLoad = (function() {
+              var link = node;
+              return function() {
+                link.removeEventListener('load', onLoad);
+                this._observeRoot(link.import);
+                this._addNodes(link.import.childNodes);
+              }.bind(this);
+            }).bind(this)();
+            if (node.import) {
+              onLoad();
+            } else {
+              node.addEventListener('load', onLoad);
+            }
+          }
+        } while (walker.nextNode())
+      }
+    },
+
+    /**
+     * @param {NodeList} nodeList
+     * @private
+     */
+    _removeNodes: function(nodeList) {
+      for (var i = 0; i < nodeList.length; i++) {
+        var root = nodeList[i];
+
+        if (!isElement(root)) {
+          continue;
+        }
+
+        // Since we're detatching this element from an observed root, we need to
+        // reobserve it.
+        // TODO(justinfagnani): can we do this in a microtask so we don't thrash
+        // on creating and destroying MutationObservers on batch DOM mutations?
+        this._observeRoot(root);
```

Comment:
Dubious that this is worth it. This is for `innerHTML`? Consider not supporting this or making it manual and/or consider patching strategy outlined in other comment.